docs/Dialects/Interop/RationaleInterop.md

```diff
@@ -0,0 +1,590 @@
+  the closest operation to perform the interop lowering and the rewrite pattern
+  driver walks top-down, thus the pattern would need to stop walking at operations
+  that also support interop, but it can only know that if it gets a list of such
+  operations or if all of the interop supporting operations implement an interface.
+* The dialect exports a function to populate a rewrite pattern that matches on
+  the interop cast operation. The disadvantage there is that we then potentially
+  have a lot of rewrite patterns matching on the same operation and we thus need
+  to specify the matching condition to only succeed when the closest interop
+  supporting operation is the one that provided the rewrite pattern. Therefore,
+  we would still need to know which operations support interop. Alternatively,
+  the pattern returns the operation it would match on and the infra tries all
+  of them and selectes the one with the closest operation returned, however,
+  that doesn't fit the dialect conversion framework very well and would be
+  rather bad performance-wise.
+* The dialect adds a trait to the operations that support interop. The interop
+  rewrite patterns could then be provided by the downstream dialect, match on
+  the unrealized interop operations, and add as a matching condition that no
+  other operation having the interop trait is closer to the unrealized interop
+  operation. This still has the disadvantage that the conversion framework as
+  to cycle through potentially many patterns matching on the same operations
+  and check quite expensive matching conditions.
+* Current implementation: The dialect implements an external model for the
+  interop operation. The lowering pass infra then provides the rewrite patterns
+  that match on the unrealized interop operations (one per op) and checks for
+  the closest parent operation that implements the interface and then calls the
+  lowering provided by the interface implementation.
+
+**Why is instance-side lowering implemented using rewrite patterns?**
+
+We also have the option to implement the lowering patterns in an interface or
+directly provide a rewrite pattern with the same trade-offs as mentioned in the
+beginning of the above question. The only advantage of the interface approach
+is the inversion of the dependencies which is not important for us since we
+implement partial lowering passes for every dialect anyways. The pattern based
+approach is more straight-forward to implement though.
+
+**Why are those interfaces implemented as external models?**
+
+Because the interface implementations need to implement a lowering pattern that
+can might have to create operations from a few different dialects, it needs to
+include all of them and link against them. This can add a big burden to some
+other user that just wants to use the dialect but is not interested in the
+interop functionality. It also helps to avoid dependency cycles.
```

Comment:
```suggestion
Because the interface implementations need to implement a lowering pattern that
might have to create operations from a few different dialects, it needs to
include all of them and link against them. This can add a big burden to some
other user that just wants to use the dialect but is not interested in the
interop functionality. It also helps to avoid dependency cycles.
```